Add option to only show favicons for pinned tabs

Ban3 opened the pull request:
Closes #3440 

Options modelled after ``tabs.show``: ``always``, ``never`` and ``pinned``.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/3680)
<!-- Reviewable:end -->

nwoeanhinnogaehr commented:
Is this not doable without any changes, just by clearing `tabs.title.format_pinned`? It sort of works but it seems that it's buggy. With that method, if I click a link opening a new tab from a pinned tab then the favicon disappears until I refresh the pinned tab.

jgkamat commented:
@nwoeanhinnogaehr as far as I know, if you clear `title.format_pinned` it should not clear the favicon, since the favicon isn't controlled by that variable. Maybe you visited a site without a favicon?

jgkamat commented on qutebrowser/config/configdata.yml:
```diff
@@ -1247,9 +1247,14 @@ tabs.favicons.scale:
-  default: true
-  type: Bool
-  desc: Show favicons in the tab bar.
+  default: always
```
This needs to have a config migration, but I'm not sure how that's handled. Just a pointer for later.

I'm not 100% sure about this, but a similar problem was solved in https://github.com/qutebrowser/qutebrowser/pull/3577 as well

The-Compiler replied:
You'd need to do something like this:

https://github.com/qutebrowser/qutebrowser/blob/8090d3e28940f02e94858d80293dcfcd8a635fd6/qutebrowser/config/configfiles.py#L251-L263

jgkamat commented on qutebrowser/mainwindow/tabbedbrowser.py:
```diff
@@ -527,7 +527,9 @@ def _get_new_tab_idx(self, related):
         """Update favicons when config was changed."""
         for i, tab in enumerate(self.widgets()):
-            if config.val.tabs.favicons.show:
+            if (config.val.tabs.favicons.show == 'always' or
```
These couple of lines are repeated several times, maybe we could make a `show_favicon` helper method either here (taking a tab) or a method in tab itself, to make this a bit cleaner?

jgkamat commented on qutebrowser/mainwindow/tabwidget.py:
```diff
@@ -108,6 +108,17 @@ def set_tab_pinned(self, tab: QWidget,
         bar.set_tab_data(idx, 'pinned', pinned)
         tab.data.pinned = pinned
+
+        if config.val.tabs.favicons.show == 'pinned':
```
I don't like that this logic is here (especially the `tabs_are_windows` part), because there's an almost exact copy in `tabbedbrowser:_update_favicons`. 

Maybe you could move the code from `_update_favicons` into a method here in tabwidget (which updates a single tabs favicons), and have the tabbedbrowser `_update_favicons` iterate over the tabs calling that method? After that, you should be able to call the tabwidget version of 'update_favicons' here.

The-Compiler replied:
I smell a conflict with #3613 :wink:

nwoeanhinnogaehr commented:
@jgkamat Ah, so it turns out gmail is weird and clears the favicon whenever you click on a link. Please disregard my previous comment.

The-Compiler commented on qutebrowser/browser/browsertab.py:
```diff
@@ -895,3 +895,7 @@ def __repr__(self):
     def is_deleted(self):
         return sip.isdeleted(self._widget)
+
+    def should_show_icon(self):
```
I try to avoid having "utility" stuff (which likely wouldn't be an useful API for future plugins) on the `BrowserTab` object. What about moving this to `TabData` as it's closely related to that anyways?

The-Compiler commented on qutebrowser/config/configfiles.py:
```diff
@@ -268,6 +268,15 @@ def _handle_migrations(self, settings):
+                if isinstance(val, bool):
+                    settings[name][scope] = 'always' if val else 'never'
+                    self._mark_changed()
+
```
Please also add tests for this to `test_configfiles.py`.

The-Compiler commented:
@Ban3 Any update on this?

Ban3 commented:
@The-Compiler Sorry, got a little busy. I did the requested changes, rebased and squashed.

I implemented `icon()` by returning the window icon in `FakeWebTab` so that the pinning tests won't fail. I didn't look too deeply into it, so I don't know if that's what you want, but it's enough to pass the tests.

Found out that `_update_favicons()` makes qutebrowser crash when playing with `tabs_are_windows`, but that also happens with master. Want a separate issue for it?

```
21:52:42 ERROR: Uncaught exception
Traceback (most recent call last):
  File "/home/user/projects/qutebrowser/qutebrowser/mainwindow/tabbedbrowser.py", line 143, in _on_config_changed
    self._update_favicons()
  File "/home/user/projects/qutebrowser/qutebrowser/mainwindow/tabbedbrowser.py", line 534, in _update_favicons
    for i, tab in enumerate(self.widgets()):
  File "/home/user/projects/qutebrowser/qutebrowser/mainwindow/tabbedbrowser.py", line 171, in widgets
    for i in range(self.widget.count()):
RuntimeError: wrapped C/C++ object of type TabWidget has been deleted
```

jgkamat commented on tests/helpers/stubs.py:
```diff
@@ -266,6 +266,9 @@ def load_status(self):
         pass
 
+    def icon(self):
+        return self.windowIcon()
```
I think it might be a bit cleaner to return a plain `QIcon()` here, but I don't think that matters too much.

Ban3 replied:
I was thinking about returning `QIcon()`, but went with `windowIcon()` because it doesn't need a extra import line. Easy to change, though.

The-Compiler replied:
I also prefer `QIcon()`, but I fixed that up in 32568a6da44468a03a2e08b1a0cbe8c4a913c3ca

jgkamat commented:
I can reproduce that crash too, I'll see if I can figure out what's going on there. If not, we definetly need to open an issue for that. (to reproduce, you set `tabs_are_windows`, open a couple windows, delete a window, and toggle the favicon setting).

The-Compiler commented:
I [fixed the crash](https://github.com/qutebrowser/qutebrowser/commit/c7e5033eaa3580d5044a3393dad39e3fb53dc418) in the latest master.

The-Compiler commented on tests/unit/config/test_configfiles.py:
```diff
@@ -233,6 +233,22 @@ def test_bindings_default(self, yaml, autoconfig):
+        if isinstance(show, bool):
+            when = 'always' if show else 'never'
+        else:
+            when = show
```
It often pays off to be a bit more explicit about expected values (and it's also nice "documentation") - I changed that (also for the test above) in 32568a6da44468a03a2e08b1a0cbe8c4a913c3ca.

The-Compiler commented:
Thanks!
